src/pelz/main.c

```diff
@@ -330,39 +340,211 @@ int main(int argc, char **argv)
+
+      uint8_t *data = NULL;
+      size_t data_len = 0;
+
+      if (read_bytes_from_file(path_id, &data, &data_len))
+      {
+        pelz_log(LOG_ERR, "seal input data file read error ... exiting");
+        free(data);
+        free(path_id);
+        free(outPath);
+        return 1;
+      }
+      pelz_log(LOG_DEBUG, "read in %d bytes of data to be wrapped", data_len);
+
+      // validate non-empty plaintext buffer specified
+      if (data_len == 0 || data == NULL)
+      {
+        pelz_log(LOG_ERR, "no input data ... exiting");
+        free(data);
+        free(path_id);
+        free(outPath);
+        return 1;
+      }
+
+      sgx_create_enclave(ENCLAVE_PATH, 0, NULL, NULL, &eid, NULL);
+
+      uint8_t *sgx_seal = NULL;
+      size_t sgx_seal_len = 0;
+      uint16_t key_policy = SGX_KEYPOLICY_MRSIGNER;
+      sgx_attributes_t attribute_mask;
+
+      attribute_mask.flags = 0;
+      attribute_mask.xfrm = 0;
+
+      if (kmyth_sgx_seal_nkl(eid, data, data_len, &sgx_seal, &sgx_seal_len, key_policy, attribute_mask))
       {
-        printf("SGX seal call not added\n");
+        pelz_log(LOG_ERR, "SGX seal failed");
+        free(data);
+        free(path_id);
+        free(outPath);
+        return 1;
       }
+
+      sgx_destroy_enclave(eid);
+      free(data);
+
+      uint8_t *tpm_seal = NULL;
+      size_t tpm_seal_len = 0;
+
       if (tpm)
       {
-        printf("Kmyth TPM call not added\n");
+        char *authString = NULL;
+        size_t auth_string_len = 0;
+        const char *ownerAuthPasswd = "";
+        size_t oa_passwd_len = 0;
+        char *cipherString = NULL;
+        int *pcrs = NULL;
+        int pcrs_len = 0;
+
+        if (tpm2_kmyth_seal(sgx_seal, sgx_seal_len, &tpm_seal, &tpm_seal_len, (uint8_t *) authString, auth_string_len,
+            (uint8_t *) ownerAuthPasswd, oa_passwd_len, pcrs, pcrs_len, cipherString))
+        {
+          pelz_log(LOG_ERR, "Kmyth TPM seal failed");
+          free(pcrs);
+          free(path_id);
+          free(outPath);
+          free(tpm_seal);
+          return 1;
+        }
+        free(pcrs);
+      }
+
+      if ((outPath != NULL) && (outPath_size != 0))
+      {
+        if (tpm)
+        {
+          if (write_bytes_to_file(outPath, tpm_seal, tpm_seal_len))
+          {
+            pelz_log(LOG_ERR, "error writing data to .ski file ... exiting");
+            free(outPath);
+            free(path_id);
+            free(tpm_seal);
+            return 1;
+          }
+          free(tpm_seal);
+        }
+        else
+        {
+          if (write_bytes_to_file(outPath, sgx_seal, sgx_seal_len))
+          {
+            pelz_log(LOG_ERR, "error writing data to .nkl file ... exiting");
+            free(outPath);
+            free(path_id);
+            free(sgx_seal);
+            return 1;
+          }
+          free(sgx_seal);
+        }
+      }
+      else
+      {
+        char ext[4];            // 4 is the length of a file extension with the period
```

Comment:
It might be cleaner to have 2 extensions as static char*'s and then just set ext based on which you need, rather than doing a memcpy.